WooCommerce/Classes/ViewRelated/Orders/Order Creation/ProductsSection/AddProductToOrderViewModel.swift

```diff
@@ -29,31 +40,155 @@ final class AddProductToOrderViewModel: ObservableObject {
+                                                       pageSize: pageSize,
+                                                       stockStatus: nil,
+                                                       productStatus: nil,
+                                                       productType: nil,
+                                                       productCategory: nil,
+                                                       sortOrder: .nameAscending,
+                                                       shouldDeleteStoredProductsOnFirstPage: false) { [weak self] result in
+            guard let self = self else { return }
+
+            switch result {
+            case .failure(let error):
+                DDLogError("⛔️ Error synchronizing products during order creation: \(error)")
+            case .success:
+                self.updateProductsResultsController()
+            }
+
+            self.transitionToResultsUpdatedState()
+            onCompletion?(result.isSuccess)
+        }
+        stores.dispatch(action)
+    }
+
+    /// Sync first page of products from remote if needed.
+    ///
+    func syncFirstPage() {
+        syncingCoordinator.synchronizeFirstPage()
+    }
+
+    /// Sync next page of products from remote.
+    ///
+    func syncNextPage() {
+        let lastIndex = productsResultsController.numberOfObjects - 1
+        syncingCoordinator.ensureNextPageIsSynchronized(lastVisibleIndex: lastIndex)
+    }
+}
+
+// MARK: - Finite State Machine Management
+private extension AddProductToOrderViewModel {
+    /// Update state for sync from remote.
+    ///
+    func transitionToSyncingState() {
+        shouldShowScrollIndicator = true
+        if products.isEmpty {
+            syncStatus = .firstPageSync
+        }
+    }
+
+    /// Update state after sync is complete.
+    ///
+    func transitionToResultsUpdatedState() {
+        shouldShowScrollIndicator = false
+        syncStatus = products.isNotEmpty ? .results: .empty
+    }
+}
+
 // MARK: - Configuration
 private extension AddProductToOrderViewModel {
+    /// Performs initial fetch from storage and updates sync status accordingly.
+    ///
     func configureProductsResultsController() {
+        updateProductsResultsController()
+        transitionToResultsUpdatedState()
+    }
+
+    /// Fetches products from storage.
+    ///
+    func updateProductsResultsController() {
         do {
             try productsResultsController.performFetch()
         } catch {
             DDLogError("⛔️ Error fetching products for new order: \(error)")
         }
     }
+
+    /// Setup: Syncing Coordinator
+    ///
+    func configureSyncingCoordinator() {
+        syncingCoordinator.delegate = self
+    }
+}
+
+// MARK: - Utils
+extension AddProductToOrderViewModel {
+    /// Represents possible statuses for syncing products
+    ///
+    enum SyncStatus {
+        case firstPageSync
+        case results
+        case empty
+        case none
```

Comment:
nit: `none` is used only as initial value. Do we really need it?

Reply:
Good point! Removed that case in 0a0dece.